tool/src/org/antlr/v4/codegen/model/InvokeRule.java

```diff
@@ -24,40 +24,43 @@
 
 /** */
 public class InvokeRule extends RuleElement implements LabeledOp {
-	public String name;
-	public OrderedHashSet<Decl> labels = new OrderedHashSet<Decl>(); // TODO: should need just 1
-	public String ctxName;
+	public final String originalName;
+	public final String name;
+	public final OrderedHashSet<Decl> labels = new OrderedHashSet<Decl>(); // TODO: should need just 1
+	public final String ctxName;
 
 	@ModelElement public List<ActionChunk> argExprsChunks;
 
 	public InvokeRule(ParserFactory factory, GrammarAST ast, GrammarAST labelAST) {
 		super(factory, ast);
 		if ( ast.atnState!=null ) {
-			RuleTransition ruleTrans = (RuleTransition)ast.atnState.transition(0);
 			stateNumber = ast.atnState.stateNumber;
 		}
 
-		this.name = ast.getText();
 		CodeGenerator gen = factory.getGenerator();
-		Rule r = factory.getGrammar().getRule(name);
-		ctxName = gen.getTarget().getRuleFunctionContextStructName(r);
+		Target target = gen.getTarget();
+		String identifier = ast.getText();
+		Rule r = factory.getGrammar().getRule(identifier);
+		this.originalName = r.name;
+		this.name = r.escapedName;
```

Comment:
Yep, the `name` should either be the original or say it is escaped.   I think this should be consistent with code generation templates. Hmm... now that I think about it I actually do prefer `r.name` as the original so that it is as it was, without change, both here and in the code generation templates.   I think it's understood that `<r.name>` is the original name and this leaves many fewer changes when we look at this change later.

Reply:
In this way, we are returning to the option when `for` is being transformed to `RULE_for_` (and this option causes other fixes, see https://github.com/antlr/antlr4/pull/3453). We've decided to escape only conflicting identifiers, i.e.: `for_`.

Reply:
Well, We are assuming `RULE_<r.name>` here since we know that there will not be a collision.